presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,185 @@
+                if (bitLengths.get(bitLengthIndex) > 0) {
+                    dimensions[dimensionIndex]++;
+                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                }
+            }
+        }
+        return dimensions;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param valueRanges the list of minimum and maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    List<ZAddressRange<Long>> zOrderCondensedSearchCurve(List<ZValueRange> valueRanges, RangeType rangeType)
+    {
+        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, addressRanges, fillInNulls(valueRanges), rangeType);
+        return addressRanges;
+    }
+
+    private List<ZValueRange> fillInNulls(List<ZValueRange> valueRanges)
+    {
+        List<ZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
+        for (int index = 0; index < valueRanges.size(); index++) {
+            ZValueRange range = valueRanges.get(index);
+
+            Integer minimumValue = range.getMinimumValue();
+            Integer maximumValue = range.getMaximumValue();
+
+            if (minimumValue == null) {
+                minimumValue = 0;
+            }
+            if (maximumValue == null) {
+                maximumValue = (1 << encodingBits.get(index)) - 1;
+            }
+
+            nonNullValues.add(new ZValueRange(Optional.of(minimumValue), Optional.of(maximumValue)));
+        }
+        return nonNullValues;
+    }
+
+    /**
+     * Recursive function to find all address ranges within given value ranges.
+     * <p/>
+     * Returns when the address range is either completely within or completely out of the value ranges.
+     * Recurses to a smaller Z-order curve if the address range only partially overlaps with the value ranges.
+     *
+     * @param level the level of recursion on the Z-curve
+     * @param startAddress the starting address of the current Z-curve level
+     * @param totalCurveBits the amount of bits in the current Z-curve level
+     */
+    private void findAddressesRecursively(int level, long startAddress, int totalCurveBits, List<ZAddressRange<Long>> addressRanges, List<ZValueRange> valueRanges, RangeType rangeType)
+    {
+        long endAddress = startAddress + ((long) 1 << totalCurveBits) - 1;
+
+        List<Integer> startValues = zOrder.decode(startAddress);
+        List<Integer> endValues = zOrder.decode(endAddress);
+
+        if (inRange(startValues, endValues, valueRanges)) {
```

Comment:
We might be able only do one pass and return `IN_RANGE`, `OUT_OF_RANGE` and `OVERLAPPING`, but if this is not on the hot path, I don't have strong opinion here.

Reply:
I've just implemented this suggestion. Compared to the previous code, this function manages to return `IN_RANGE`, `OUT_OF_RANGE` and `OVERLAPPING` in one loop.
However, I'm not sure how combining this logic affects the readability of the method.
What do you think?

Reply:
As discussed offline, the potential performance gain is not a lot given it is small loop outside of code path, but the readability loss is significant. It might not be a good trade-off. Thanks a lot for prototyping it :)